src/ir/auth_logic/ast.h

```diff
@@ -93,32 +93,25 @@ class BaseFact {
 class CanSay;
 class Fact;
 
-// CanSay corresponds to an expression of the form <principal> canSay Fact
 class CanSay {
  public:
-  explicit CanSay(Principal principal, std::unique_ptr<Fact>& fact)
-      : principal_(principal), fact_(std::move(fact)) {}
+  explicit CanSay(Principal principal) : principal_(std::move(principal)) {}
   const Principal& principal() const { return principal_; }
-  const Fact* fact() const { return fact_.get(); }
 
  private:
   Principal principal_;
-  std::unique_ptr<Fact> fact_;
 };
 
-// Fact corresponds to either a base fact or a an expression of the form
-// <principal> canSay <Fact>
 class Fact {
  public:
-  // FactVariantType gives the different forms of Facts. Client code
-  // should use this type to traverse these forms. This type may be changed in
-  // the future.
-  using FactVariantType = std::variant<BaseFact, std::unique_ptr<CanSay>>;
-  explicit Fact(FactVariantType value) : value_(std::move(value)) {}
-  const FactVariantType& GetValue() const { return value_; }
+  Fact(std::forward_list<CanSay> can_say, BaseFact base_fact)
+      : can_say_(std::move(can_say)), base_fact_(std::move(base_fact)) {}
+  const std::forward_list<CanSay>& can_say() const { return can_say_; }
+  const BaseFact& base_fact() const { return base_fact_; }
 
  private:
-  FactVariantType value_;
+  std::forward_list<CanSay> can_say_;
```

Comment:
If you get rid of `CanSay`, then you can change this to `std::forward_list<Principal> delegatees_;` or `std::forward_list<Principal> delegation_chain_;`?

Reply:
Done!